### DISCUSSION TITLE
Images built with docker-compose (via `podman compose`) are assigned to `docker.io/library` registry

### DISCUSSION THREAD

**Discussion by xduugu** [containers/podman#24445] [General]
### Issue Description

When building an image with docker-compose, its name is prefixed with the `docker.io/library` registry name instead of `localhost`.

### Steps to reproduce the issue

Steps to reproduce the issue
1. Install `docker-compose`, so that it is usable by `podman compose`
2. Create a `Dockerfile` file with content:
```Dockerfile
FROM scratch
```
3. Create a compose file `compose.yaml` in the same directory with content:
```yaml
name: test-project
services:
  test-service:
    build: .
```
4. Run `podman compose build` in this directory

### Describe the results you received

A new image was created with name `docker.io/library/test-project-test-service`:

```
Sending build context to Docker daemon     372B
STEP 1/2: FROM scratch
STEP 2/2: LABEL "com.docker.compose.image.builder"="classic"
COMMIT docker.io/library/test-project-test-service
--> 635e9c6bae57
Successfully tagged docker.io/library/test-project-test-service:latest
635e9c6bae576c4cda064f8ed705a55762cbc0e0e28432c5980a33de5d56db56
Successfully built 635e9c6bae57
Successfully tagged test-project-test-service
```

### Describe the results you expected

The created image should use the name `localhost/test-project-test-service`.

### podman info output

```
host:
  arch: amd64
  buildahVersion: 1.37.5
  cgroupControllers:
  - cpu
  - io
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.12-3.fc41.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.12, commit: '
  cpuUtilization:
    idlePercent: 92.42
    systemPercent: 2.55
    userPercent: 5.03
  cpus: 16
  databaseBackend: sqlite
  distribution:
    distribution: fedora
    variant: workstation
    version: "41"
  eventLogger: journald
  freeLocks: 2048
  hostname: fedora
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 6.11.5-300.fc41.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 16027574272
  memTotal: 32822956032
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns-1.12.2-2.fc41.x86_64
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.12.2
    package: netavark-1.12.2-1.fc41.x86_64
    path: /usr/libexec/podman/netavark
    version: netavark 1.12.2
  ociRuntime:
    name: crun
    package: crun-1.18-1.fc41.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.18
      commit: 8656b2548509fcc69ea7e8823a870564360a57a1
      rundir: /run/user/1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20240906.g6b38f07-1.fc41.x86_64
    version: |
      pasta 0^20240906.g6b38f07-1.fc41.x86_64
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: false
    path: /run/user/1000/podman/podman.sock
  rootlessNetworkCmd: pasta
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: false
  serviceIsRemote: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 8589930496
  swapTotal: 8589930496
  uptime: 1h 10m 0.00s (Approximately 0.04 days)
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
store:
  configFile: /home/xduugu/.config/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/xduugu/.local/share/containers/storage
  graphRootAllocated: 998380109824
  graphRootUsed: 187039444992
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "false"
    Supports volatile: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 0
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/xduugu/.local/share/containers/storage/volumes
version:
  APIVersion: 5.2.5
  Built: 1729209600
  BuiltTime: Fri Oct 18 02:00:00 2024
  GitCommit: ""
  GoVersion: go1.23.2
  Os: linux
  OsArch: linux/amd64
  Version: 5.2.5
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

### Additional environment details

Docker Compose version v2.30.1

### Additional information

**Comment by baude** [comment]
I *think* this is an assumption made by docker compose code itself ?  I would need to verify that but I have a vague recollection about this.

**Comment by Luap99** [comment]
I am pretty sure this is enforced by the compat API to match docker behavior.

You can set compat_api_enforce_docker_hub=false in containers.conf to disable this AFAIK.
https://github.com/containers/common/blob/main/docs/containers.conf.5.md

**Reply by vrothberg** [reply]
@Luap99 is spot on.  Requests going through the Docker REST API always resolve short names to docker.io.  That is to remain compatible with Docker.

The linked option in containers.conf can disable this behavior.  It must be configured on the server side.  So if you're running against a podman machine, it must be configured in that VM.

**Reply by xduugu** [reply]
Thanks a lot, @Luap99 and @vrothberg. It would never have found this config option on my own.

However, I don't really understand the need for this option. Are there any drawbacks when setting it to `false`? When I worked with docker several years ago, I never saw the full qualified image names, but it could be that all images used the `docker.io` prefix internally.

Or is it related to the fact that `localhost/` images are not considered for unqualified image names (#23055)?

**Reply by vrothberg** [reply]
> However, I don't really understand the need for this option.

It exists for historic reasons.  Initially, the Docker-compat API used localhost/ but it to be really compatible it needed to resolve to docker.io.  To enable both scenarios, we added the config.

> Are there any drawbacks when setting it to false? When I worked with docker several years ago, I never saw the full qualified image names, but it could be that all images used the docker.io prefix internally.

Some tools expect short names to resolve to docker.io.  Most use cases don't care though.

**Reply by Luap99** [reply]
The history/reasons are well documented in the git log/github issues
https://github.com/containers/podman/commit/5bdd571b1e46f26e23f030456efb009cbb765e4c
https://github.com/containers/podman/issues/12320

**Reply by xduugu** [reply]
Thanks again, @vrothberg and @Luap99, for the background.

I totally understand the behavior when pulling unqualified images, but imo it's unfortunate that podman assigns names to locally built images that possibly conflict with actual images on dockerhub.

For example, if I use this `compose.yaml`, the resulting image name will be `docker.io/library/mongo-express`, which conflicts with [mongo-express](https://hub.docker.com/_/mongo-express) on dockerhub.

```yaml
name: mongo
services:
  express:
    build: .
```

If the dockerhub image had been pulled before, it will get untagged and the name is assigned to the locally built image. If the the dockerhub image is then pulled, the locally built image gets untagged.

By the way, I noticed that if there already is an image with the name `localhost/mongo-express`, the REST api will use this name instead of the one with the `docker.io/library/` prefix. I guess that's because of #14294.

It's fine for me if you say that's actually the intended behavior, especially since there is a configuration option, but it seems strange to me.

**Reply by rhatdan** [reply]
What would Docker do in this situation?

**Reply by vrothberg** [reply]
> What would Docker do in this situation?

Docker only resolves to "docker.io", so it would ignore "localhost/foo" when looking up "foo" in the local storage.  Strictly speaking, the compat API does not behave correctly here.
